src/Authenticator/JwtAuthenticator.php

```diff
@@ -0,0 +1,138 @@
+use Psr\Http\Message\ServerRequestInterface;
+use RuntimeException;
+use stdClass;
+
+class JwtAuthenticator extends TokenAuthenticator {
+
+    /**
+     * @inheritDoc
+     */
+    protected $_defaultConfig = [
+        'header' => 'Authorization',
+        'queryParam' => 'token',
+        'tokenPrefix' => 'bearer',
+        'allowedAlgs' => ['HS256'],
+        'entityClass' => Entity::class,
+        'returnPayload' => true,
+        'key' => null,
+        'salt' => null
+    ];
+
+    /**
+     * Payload data.
+     *
+     * @var object|null
+     */
+    protected $_payload;
+
+    /**
+     * @inheritdoc
+     */
+    public function __construct(IdentifierCollection $identifiers, array $config = [])
+    {
+        parent::__construct($identifiers, $config);
+
+        if (empty($this->_config['salt'])) {
+            if (class_exists('\Cake\Utility\Security')) {
+                $this->getConfig('salt', \Cake\Utility\Security::salt());
+            } else {
+                throw new RuntimeException('You must set the `salt` config key');
+            }
+        }
+    }
+
+    /**
+     * Authenticates the identity contained in a request. Will use the `config.userModel`, and `config.fields`
+     * to find POST data that is used to find a matching record in the `config.userModel`. Will return false if
+     * there is no post data, either username or password is missing, or if the scope conditions have not been met.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request that contains login information.
+     * @param \Psr\Http\Message\ResponseInterface $response Unused response object.
+     * @return \Authentication\ResultInterface
+     */
+    public function authenticate(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        $result = $this->getPayload($request);
+
+        if (!$result instanceof stdClass) {
+            return new Result(null, Result::FAILURE_CREDENTIAL_INVALID);
+        }
+
+        $result = json_decode(json_encode($result), true);
+
+        if (empty($result)) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_NOT_FOUND);
+        }
+
+        if ($this->getConfig('returnPayload')) {
+            $entityClass = $this->getConfig('entityClass');
+            $entity = new $entityClass($result);
+
+            return new Result($entity, Result::SUCCESS);
+        }
+
+        $user = $this->identifiers()->identify($result);
+
+        if (empty($user)) {
+            return new Result(null, Result::FAILURE_IDENTITY_NOT_FOUND, $this->identifiers()->getErrors());
+        }
+
+        return new Result($user, Result::SUCCESS);
+    }
+
+    /**
+     * Get payload data.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface|null $request Request to get authentication information from.
+     * @return object|null Payload object on success, null on failure
+     */
+    public function getPayload(ServerRequestInterface $request = null)
```

Comment:
It's so that one can get the payload after authentication has been completed. JWT can contain permission info too, so authorizers would have need access to the payload. Once payload is already decoded from the request, the request object is not required again.